x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1800,20 +1815,19 @@ static BytesReference createTokenDocument(
                     .field("type", "unassociated_client")
                     .field("user", originatingClientAuth.getUser().principal())
                     .field("realm", originatingClientAuth.getAuthenticatedBy().getName());
-                if (originatingClientAuth.getSourceRealm().getDomain() != null) {
-                    builder.field("realm_domain", originatingClientAuth.getSourceRealm().getDomain());
+                if (originatingClientAuth.getAuthenticatedBy().getDomain() != null) {
+                    builder.field("realm_domain", originatingClientAuth.getAuthenticatedBy().getDomain());
                 }
```

Comment:
The client's authentication details that are stored in the token document, in the code above, are used only when verifying that the client refreshing the token is "the same" to the client that created the token.
In order to use the existing code in `Authentication#canAccessResourcesOf` for the same user check, we would need to store the serialized authentication, instead of only the picked fields.
What do you think, should we store the serialized authentication instead, in order to reuse that code?
I think we should because the intent is that the refresh token be owned, the same way that other things are owned.

If not, the "same user" check also inspects the realm type in addition to the realm name.
Should we record the realm type in order to perform an equivalent sort of access verification?

Reply:
I think that is likely what we want to do. But we need discussion about it because it has subtle "breakage". The current logics compares the tuple of "effective username" and "authentication realm name". This is a cross between authentication user and effective user which is not consistent to any other use cases. I think it is a bug. Once we decided that it should be fixed. We might as well decide to reuse the `canAccessResourcesOf` logic for it (in a separate PR).